ydb/core/log_backend/json_envelope.h

```diff
@@ -0,0 +1,48 @@
+#pragma once
+#include <library/cpp/json/json_reader.h>
+#include <library/cpp/json/json_writer.h>
+
+#include <util/generic/string.h>
+
+#include <vector>
+
+namespace NKikimr {
+
+class TJsonEnvelope {
+public:
+    explicit TJsonEnvelope(const TString& templateString)
+        : TemplateString(templateString)
+    {
+        Parse(); // can throw
+    }
+
+    TJsonEnvelope() = delete;
+    TJsonEnvelope(const TJsonEnvelope&) = delete;
+    TJsonEnvelope(TJsonEnvelope&&) = delete;
+
+    TString ApplyJsonEnvelope(const TStringBuf& message);
+
+private:
+    void Parse();
+    void Parse(NJson::TJsonValue* value);
+
+private:
+    struct TReplace {
+        NJson::TJsonValue* Value = nullptr;
+        std::vector<TString> ReplaceSequence; // empty string for placeholder
+
+        TReplace(NJson::TJsonValue* value)
+            : Value(value)
+        {}
+
+        bool Parse(const TString& replace);
+        void Apply(const TStringBuf& message);
+    };
+
+private:
+    TString TemplateString;
+    NJson::TJsonValue Value;
+    std::vector<TReplace> Replaces;
```

Comment:
Is there a reason to support multiple placeholder replacements?
Enveloping in general assumes wrapping around a single payload. And single replacement would have made template rendering code much simpler.

Reply:
No, there is no explicit requirements for such replacement. I can rewrite the code and make it simpler.